feat: MMPD-1528 - View recents and favorites in browser URL modal

ziad-saab opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**
https://consensyssoftware.atlassian.net/browse/MMPD-1528

Change the behaviour of the browser url autocomplete:

- add scrolling
- when nothing is typed, show recents (max 5) and favorites
- when user starts typing, auto complete from dapp list + recents + favorites
- adjust styling to match new figma styles

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/MMPD-1528

## **Manual testing steps**

1. Go to the in-app browser
1. Press the URL bar
1. Observe that the currently typed URL is available as a result, just like in the current version
1. Clear the input and observe that you see your (up to) 5 latest recents, + favorites, and that you can scroll
1. Start typing, and observe that you get up to 3 lists: Sites, Recents, and Favorites. The elements shown are matched from what you're typing

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<img width="428" alt="image" src="https://github.com/user-attachments/assets/e92ca642-ba5d-4391-9995-76bdb9b38f57" />

<img width="434" alt="image" src="https://github.com/user-attachments/assets/dfa0daa0-a978-4035-b1c4-0340273c16e7" />

<img width="408" alt="image" src="https://github.com/user-attachments/assets/3aa3e1fe-8dc7-4f30-b4b0-3dd9e1ae7271" />

<img width="400" alt="image" src="https://github.com/user-attachments/assets/839afc18-c117-452c-b0bd-1e74f47fdcef" />


https://github.com/user-attachments/assets/0f0c46eb-e752-495f-a7b7-f0fdb1d0d5d7





## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

ziad-saab committed:
feat: change browser url modal behavior

- when nothing is typed, show recents (max 5) and favorites
- when user starts typing, show results from the dapp list

ziad-saab committed:
Merge branch 'main' into zs/new-browser-autocomplete

ziad-saab committed:
update snapshots

ziad-saab committed:
Autocomplete on recents and favorites

ziad-saab committed:
limit recents to 5 results

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/new-browser-autocomplete

ziad-saab committed:
re-set search wrapper new background color

ziad-saab committed:
adjust size of no-results autocomplete container

ziad-saab committed:
update snapshots

ziad-saab committed:
linting

ziad-saab committed:
increase test coverage of UrlAutocomplete

ziad-saab committed:
linting

ziad-saab committed:
Merge branch 'main' into zs/new-browser-autocomplete

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/new-browser-autocomplete

ziad-saab committed:
Add trash icon to remove favorites

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/new-browser-autocomplete

ziad-saab committed:
remove duplicate import

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/new-browser-autocomplete

ziad-saab committed:
remove commented code

ziad-saab committed:
Merge branch 'main' into zs/new-browser-autocomplete

Cal-L commented on app/components/UI/UrlAutocomplete/index.js:
```diff
@@ -14,14 +15,27 @@ import WebsiteIcon from '../WebsiteIcon';
+const MAX_RECENTS = 5;
+const ORDERED_CATEGORIES = ['sites', 'recents', 'favorites'];
+
+export const deleteFavoriteTestId = (url) => `delete-favorite-${url}`;
```
Move this into a .testIds file under `testIDs/BrowserScreen`

Cal-L commented on app/components/UI/UrlAutocomplete/index.js:
```diff
@@ -14,14 +15,27 @@ import WebsiteIcon from '../WebsiteIcon';
+import { IconName } from '../../../component-library/components/Icons/Icon';
+import { removeBookmark } from '../../../actions/bookmarks';
+
+const MAX_RECENTS = 5;
```
Move these two into a constants file in this directory

Cal-L commented on app/components/UI/UrlAutocomplete/index.js:
```diff
@@ -36,6 +50,12 @@ const createStyles = (colors) =>
+    category: {
+      fontSize: 16,
+      color: colors.text.default,
+      ...fontStyles.bold,
```
Use typography from theme based on what font style Figma is showing. For example, reference `Input.styles.ts`

ziad-saab replied:
@Cal-L need more details on this one please

Cal-L commented on app/components/UI/UrlAutocomplete/index.js:
```diff
@@ -180,20 +222,30 @@ class UrlAutocomplete extends PureComponent {
-      );
-    if (this.state.results.length === 0) {
+    let results = [];
+    if (!this.props.input || this.props.input.length === 0) {
```
Wrap this logic in a useMemo since it will recreate the array and call reverse every time the component re-renders

Cal-L commented on app/components/UI/UrlAutocomplete/index.js:
```diff
@@ -180,20 +222,30 @@ class UrlAutocomplete extends PureComponent {
+      results = this.state.results;
+    }
+
+    const resultsByCategory = results.reduce((acc, currentResult) => {
```
Wrap this in a useMemo to optimize for performance

Cal-L commented on app/components/UI/UrlAutocomplete/index.js:
```diff
@@ -215,30 +267,45 @@ class UrlAutocomplete extends PureComponent {
       );
     }
+
+    const categoriesWithResults = ORDERED_CATEGORIES.filter(category => resultsByCategory[category]?.length > 0);
```
Wrap this in a useMemo to optimize for performance

Cal-L commented on app/components/UI/UrlAutocomplete/index.js:
```diff
@@ -215,30 +267,45 @@ class UrlAutocomplete extends PureComponent {
-        })}
+      <ScrollView style={styles.wrapper} contentContainerStyle={styles.contentContainer}>
+        {
+          categoriesWithResults.map(category => (
```
Abstract this into a render function and wrap it in useCallback to optimize for performance

Cal-L commented on app/components/UI/UrlAutocomplete/index.js:
```diff
@@ -215,30 +267,45 @@ class UrlAutocomplete extends PureComponent {
 
 const mapStateToProps = (state) => ({
-  browserHistory: state.browser.history,
+  browserHistory: state.browser.history.map(i => ({...i, type: 'recents'})),
```
I'd recommend abstracting these into selectors that use `createDeepEqualSelector` so that it only updates on changed items instead of every time. Same for the selector below. You can create a `browser` file in `app/selectors` to store these

Cal-L commented on app/components/Views/BrowserUrlModal/styles.ts:
```diff
@@ -39,8 +40,11 @@ export const createStyles = (colors: any) =>
+      borderRadius: 8,
+      borderWidth: 1,
+      borderColor: colors.border.default,
+      height: Device.isAndroid() ? 50 : 40,
```
Is it necessary to have different heights for the platforms or can we just use one value?

ziad-saab replied:
turns out it wasn't. this was a leftover from the previous version before the save view area changes.

Cal-L reviewed (commented):
Left some comments

Cal-L reviewed (commented):
Heads up that the many use of array methods in the `UrlAutocomplete` may potentially cause performance issues. It seems that there could be a more optimal way to store the values to reduce the amount of transformations. Happy to pair on this

ziad-saab committed:
address PR feedback

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/new-browser-autocomplete

Cal-L commented on app/components/UI/UrlAutocomplete/index.js:
```diff
@@ -268,24 +159,18 @@ export class UrlAutocomplete extends PureComponent {
-    const categoriesWithResults = ORDERED_CATEGORIES.filter(category => resultsByCategory[category]?.length > 0);
-
     return (
       <ScrollView style={styles.wrapper} contentContainerStyle={styles.contentContainer}>
```
This should be updated to be a Flatlist instead of a Scrollview since it abstracts the need to map over the results.

Cal-L reviewed (commented):
Left a comment

ziad-saab committed:
address more pr feedback

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/new-browser-autocomplete

ziad-saab committed:
change scrollview > sectionlist

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/new-browser-autocomplete

ziad-saab committed:
linting

ziad-saab committed:
Merge branch 'main' into zs/new-browser-autocomplete

ziad-saab committed:
remove conditional height

ziad-saab committed:
deduplicate URL results

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/new-browser-autocomplete

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/new-browser-autocomplete

metamaskbot committed:
Bump version number to 1526

metamaskbot committed:
Bump version number to 1527

ziad-saab committed:
bump versions for testflight release

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/new-browser-autocomplete

Cal-L commented:
Are the e2e failures on your end or on main?

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/new-browser-autocomplete

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/new-browser-autocomplete

Cal-L reviewed (dismissed):
LGTM!

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/new-browser-autocomplete

ziad-saab committed:
Merge mobile dapp list with portfolio dapp list

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/new-browser-autocomplete

ziad-saab committed:
Add keyboard avoiding view to the browser url autocomplete

now that the suggestion list is scrollable, the fact that the view isn't keyboard avoiding is visible

ziad-saab committed:
Merge remote-tracking branch 'origin/main' into zs/new-browser-autocomplete

TODO:
- Port changes from UrlAutocomplete/index.temp.js into the new tsx

ziad-saab committed:
merge old autocomplete update into new version

ziad-saab committed:
remove duplicate state

ziad-saab committed:
update snapshots

ziad-saab committed:
add Keyboard avoiding view to browser tab

ziad-saab committed:
Merge branch 'main' into zs/new-browser-autocomplete

Cal-L reviewed (dismissed):
Dupe in sonarcloud exists pre changes. LGTM

ziad-saab committed:
fix tapping on list item not going to URL

Cal-L reviewed (approved):
LGTM!
